tests/test_server_client.py

```diff
@@ -0,0 +1,246 @@
+        self.conn_handle = None
+    
+    def stop(self):
+        if self.conn_handle:
+            self.adapter.driver.ble_gap_disconnect(self.conn_handle)
+
+    def start(self, connect_with):
+        self.connect_with = connect_with
+        logger.info(f"scan_start, trying to find {self.connect_with}")
+        scan_duration = 5
+        self.adapter.driver.ble_gap_scan_start()#scan_params=params) 
+        try:
+            self.conn_handle = self.conn_q.get(timeout=scan_duration)
+            self.adapter.service_discovery(self.conn_handle)
+            self.adapter.enable_notification(self.conn_handle, BLEUUID(HeartRateCharacterstic))
+            logger.info(f"Notification enabled.")
+        except Empty:
+            logger.info(f"No heart rate collector advertising with name {self.connect_with} found.")
+
+    def on_gap_evt_connected(
+        self, ble_driver, conn_handle, peer_addr, role, conn_params
+    ):
+        self.conn_q.put(conn_handle)
+        logger.info(f"(Central) New connection: {conn_handle}.")
+
+    def on_gap_evt_disconnected(self, ble_driver, conn_handle, reason):
+        logger.info(f"Disconnected: {conn_handle} {reason}.")
+
+    def on_gap_evt_adv_report(
+        self, ble_driver, conn_handle, peer_addr, rssi, adv_type, adv_data
+    ):
+        if BLEAdvData.Types.complete_local_name in adv_data.records:
+            dev_name_list = adv_data.records[
+                BLEAdvData.Types.complete_local_name
+            ]
+
+        elif BLEAdvData.Types.short_local_name in adv_data.records:
+            dev_name_list = adv_data.records[BLEAdvData.Types.short_local_name]
+        else:
+            return
+
+        dev_name = "".join(chr(e) for e in dev_name_list)
+
+        if dev_name == self.connect_with:
+            address_string = "".join(
+                "{0:02X}".format(b) for b in peer_addr.addr
+            )
+            logger.info(
+                "Trying to connect to peripheral advertising as %s, address: 0x%s",
+                dev_name,
+                address_string,
+            )
+            self.adapter.connect(peer_addr, tag=1)
+
+    def on_notification(self, ble_adapter, conn_handle, uuid, data):
+        data = data[0]
+        logger.info(f"Connection: {conn_handle}, {uuid} = {data}.")
+        self.notification_q.put(data)
+
+class Peripheral(BLEDriverObserver, BLEAdapterObserver):
+    def __init__(self, adapter):
+        self.adapter = adapter
+        logger.info(
+            "Peripheral adapter is %d",
+            self.adapter.driver.rpc_adapter.internal,
+        )
+        self.conn_q = Queue()
+        self.adapter.observer_register(self)
+        self.adapter.driver.observer_register(self)
+        self.heart_rate = random.randint(60, 200)
+        self.char_handles = None
+
+    def start(self, adv_name):
+        adv_data = BLEAdvData(complete_local_name=adv_name)
+        self.adapter.driver.ble_gap_adv_data_set(adv_data)
```

Comment:
Could split out server setup to a separate function, and call it before call to start().